[Cloud.gov] Client side Password Validation

jcscottiii opened the pull request:
Work card:
https://trello.com/c/tc2CAVkh/439-ui-changing-setting-password-should-show-requirements-beforehand

Commit:
[Cloud.gov] Client side Password Validation

Work card:
https://trello.com/c/tc2CAVkh/439-ui-changing-setting-password-should-show-requirements-beforehand

Commit:
Client side Password Validation pt.2

Adds validation for scenarios:
Activating a new account.
Accepting an invite for a new account

Work card:
https://trello.com/c/tc2CAVkh/439-ui-changing-setting-password-should-show-requirements-beforehand

msecret commented on login/src/main/resources/templates/web/accounts/new_activation_email.html:
```diff
@@ -3,7 +3,27 @@
+</head>
 <div class="island-landscape" layout:fragment="page-content">
+    <script th:inline="javascript">
+    /*<![CDATA[*/
```
I've never seen this needed when dumping into a script tag. I'm guessing this is something strange with this java templating?

jcscottiii replied:
yeah, blame thymeleaf for this.

msecret commented on login/src/main/resources/templates/web/accounts/new_activation_email.html:
```diff
@@ -3,7 +3,27 @@
+    $(document).ready(function(){
+        // Get all the password policy requirements.
+        // In case the policy equals to null, initialize the values to 0.
+        var special_count = /*[[${passwordPolicy == null ? 0 : passwordPolicy.getRequireSpecialCharacter()}]]*/;
```
These should probably be camelCased as it follows the rest of the cloud.gov standard. Unless there's a reason they shouldn't be here?

msecret commented on login/src/main/resources/templates/web/change_password.html:
```diff
@@ -4,11 +4,28 @@
     <th:block layout:include="nav :: nav"></th:block>
 </div>
 <div class="island" layout:fragment="page-content">
+  <script th:inline="javascript">
```
This may not be possible in this system, but this code shoudln't be duplicated. Like it should be imported from one location. If this is hard to do within this java system though, I'm fine leaving it and having a TODO note.

jcscottiii replied:
I'm not sure it can be. Depending on when the template library injects `passwordPolicy`. But let me try.

Commit:
Client side Password Validation pt.3

Clean up and fix code style.
Abstract redundant code
Work card:
https://trello.com/c/tc2CAVkh/439-ui-changing-setting-password-should-show-requirements-beforehand

geramirez commented on uaa/src/main/webapp/resources/javascripts/pw_validation.js:
```diff
@@ -0,0 +1,106 @@
+    document.getElementById("length-count").innerHTML=''+lengthCount+'';
+
+    // Hide rules that aren't set or set to zero.
+    if ( specialCount === 0 ) {$('#special-req').hide();}
```
Will rules that aren't set always be 0?

jcscottiii replied:
These will always be 0 in the case of "no policy" for the policies

geramirez commented on uaa/src/main/webapp/resources/javascripts/pw_validation.js:
```diff
@@ -0,0 +1,106 @@
+    document.getElementById("uppercase-count").innerHTML=''+uppercaseCount+'';
+    document.getElementById("lowercase-count").innerHTML=''+lowercaseCount+'';
+    document.getElementById("number-count").innerHTML=''+numberCount+'';
+    document.getElementById("length-count").innerHTML=''+lengthCount+'';
```
Are the empty string for casting or formatting?

jcscottiii replied:
Ensuring these are strings

geramirez commented:
:+1: awesome new feature!

msecret commented:
Looks good @jcscottiii

jcscottiii commented:
Needs #2
